services/structureddate/structureddate/src/main/java/org/collectionspace/services/structureddate/antlr/ANTLRStructuredDateEvaluator.java

```diff
@@ -460,12 +467,75 @@ public void exitInvStrDate(InvStrDateContext ctx) {
 		Integer dayOfMonth = (Integer) stack.pop();
 		Integer numMonth = (Integer) stack.pop();
 		Integer year = (Integer) stack.pop();
-		Era era = (Era) stack.pop();
+		Era era = (ctx.era() == null) ? null : (Era) stack.pop();
+
+		stack.push(year);
+		stack.push(numMonth);
+		stack.push(dayOfMonth);
+		stack.push(era);
+	}
+
+	@Override
+	public void exitDayFirstDate(DayFirstDateContext ctx) {
+		if (ctx.exception != null) return ;
+		
+		Era era = (ctx.era() == null) ? null : (Era) stack.pop();
+		Integer year = (Integer) stack.pop();
+		Integer month = (Integer) stack.pop();
+		Integer dayOfMonth = (Integer) stack.pop();
+
+		stack.push(year);
+		stack.push(month);
+		stack.push(dayOfMonth);
+		stack.push(era);
+	}
+
+	@Override
+	public void exitDayOrYearFirstDate(DayOrYearFirstDateContext ctx) {
+		if (ctx.exception != null) return;
+
+		Era era = (stack.size() == 3) ? null : (Era) stack.pop();
+		Integer year = (Integer) stack.pop();
+		Integer numMonth = (Integer) stack.pop();
+		Integer dayOfMonth = (Integer) stack.pop();
+
+		Integer num1 = year;
+		Integer num2 = dayOfMonth;
+
+		if (DateUtils.isValidDate(num1, numMonth, num2, era)) {
+			// Do nothing, it is already In the right format
+		} else if (DateUtils.isValidDate(num2, numMonth, num1, era)) {
+			// Then the first number is a year. Fix:
+			year = num2;
+			dayOfMonth = num1;
+		}
 
```

Comment:
This all works, but I find it confusing. I would make the names of the args you pop off the stack look like the names of the tokens in your grammar. Also the second number is at the top of the stack, isn't it? So something like:

```
		Integer num2 = (Integer) stack.pop();
		Integer numMonth = (Integer) stack.pop();
		Integer num1 = (Integer) stack.pop();

		Integer year;
		Integer dayOfMonth;

 		if (DateUtils.isValidDate(num1, numMonth, num2, era)) {
			// The first number is a year.
			year = num1;
			dayOfMonth = num2;
		} else if (DateUtils.isValidDate(num2, numMonth, num1, era)) {
			// The second number is a year.
			year = num2;
			dayOfMonth = num1;
		}
 ```

Reply:
Also could you add a test for an ambiguous case, e.g. "13 April 15". We should have a test for how we're interpreting that, so that it doesn't get changed without us knowing.